// libfs/libfs.js
exports.calculateDirSize = require('./lib/calculateDirSize').calculateDirSize;
exports.copyDir = require('./lib/copyDir').copyDir;
exports.copyFile = require('./lib/copyFile').copyFile;
exports.createDir = require('./lib/createDir').createDir;
exports.createFile = require('./lib/createFile').createFile;
exports.deleteDir = require('./lib/deleteDir').deleteDir;
exports.deleteFile = require('./lib/deleteFile').deleteFile;
exports.getLogicalDrives = require('./lib/getLogicalDrives').getLogicalDrives;
exports.moveDir = require('./lib/moveDir').moveDir;
exports.moveFile = require('./lib/moveFile').moveFile;
exports.queryDir = require('./lib/queryDir').queryDir;
exports.renameDir = require('./lib/renameDir').renameDir;
exports.renameFile = require('./lib/renameFile').renameFile;
exports.retriveDirInfo = require('./lib/retriveDirInfo').retriveDirInfo;
exports.retriveFileInfo = require('./lib/retriveFileInfo').retriveFileInfo;
exports.search = require('./lib/search').search;
exports.updateDirInfo = require('./lib/updateDirInfo').updateDirInfo;
exports.updateFileInfo = require('./lib/updateFileInfo').updateFileInfo;
exports.isWindows = require('./lib/isWindows').isWindows;
